libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt

```diff
@@ -82,21 +77,35 @@ class MapboxRouteLineView(var options: MapboxRouteLineOptions) {
                     RouteConstants.ALTERNATIVE_ROUTE2_SOURCE_ID,
                     routeDrawData.value.alternativeRoute2Source
                 )
-                updateLineGradient(
-                    style,
-                    RouteLayerConstants.ALTERNATIVE_ROUTE1_TRAFFIC_LAYER_ID,
-                    routeDrawData.value.altRoute1TrafficExpression
-                )
-                updateLineGradient(
-                    style,
-                    RouteLayerConstants.ALTERNATIVE_ROUTE2_TRAFFIC_LAYER_ID,
-                    routeDrawData.value.altRoute2TrafficExpression
-                )
                 updateSource(
                     style,
                     RouteConstants.WAYPOINT_SOURCE_ID,
                     routeDrawData.value.waypointsSource
                 )
+
+                routeDrawData.value.trafficLineExpressionProvider?.apply {
+                    updateLineGradient(
+                        style,
+                        RouteLayerConstants.PRIMARY_ROUTE_TRAFFIC_LAYER_ID,
+                        this()
```

Comment:
I believe that we're actually regressing here. The invocation of this function (that prepares traffic expression) now happens synchronously. This means that the time to render the route is exactly the same, with a side effect that we are freezing the main thread for an operation that previously was done asynchronously.

To achieve the goal described in the PR the expression provider functions should be executed asynchronously and then the results should be applied to map sources. We also need to remember about correct synchronization - if a new render call comes in, the traffic layers update should be canceled so that it doesn't compete with the new one. Or they should be synchronized with a mutex.

Reply:
Added a cancel similar to the vanishing route line class.